Docker documentation for v4 (#1144)
strapi/documentation: docs/developer-docs/latest/setup-deployment-guides/installation/docker.md

---
title: Custom Docker Container - Strapi Developer Docs
description: Quickly create a Docker container from a local project.
canonicalUrl: https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/installation/docker.html
---

# Running Strapi in a Docker container

:::caution
Strapi does not build any official container images. The following instructions are provided as a courtesy to the community. If you have any questions, please reach out to the community on [Discord](https://discord.strapi.io).
:::

The following documentation will guide you through the process of building a custom [Docker](https://www.docker.com/) container with an existing Strapi project.

> Docker is an open platform that allows to develop, ship and run applications by using containers (i.e. packages containing all the parts an application needs to function, such as libraries and dependencies). Containers are isolated from each other and bundle their own software, libraries and configuration files; they can communicate with each other through well-defined channels.

Reviewer: ```suggestion
Docker is an open platform that allows developing, shipping, and running applications by using containers (i.e. packages containing all the parts an application needs to function, such as libraries and dependencies). Containers are isolated from each other and bundle their own software, libraries, and configuration files; they can communicate with each other through well-defined channels.
```

> ## Prerequisites

Reviewer: WDYT about changing this to the prerequisites callout and then deleting line 19?

To follow the instructions below, the following will be needed:

- [Docker](https://www.docker.com/) installed on your machine
- [Supported version of Node.js](/developer-docs/latest/setup-deployment-guides/installation/cli.md#step-1-make-sure-requirements-are-met)
- An existing Strapi v4 project, or a new one created with [Strapi Quick Start](/developer-docs/latest/getting-started/quick-start.md)

> - (Optional) [Yarn](https://yarnpkg.com/) installed on your machine

Reviewer: I think we are using lowercase (optional) elsewhere in the docs. Maybe @pwizla can confirm?

Author: which are you referring to here?

> - (Optional) [Yarn](https://yarnpkg.com/) installed on your machine

Reviewer 2: ```suggestion
- (_optional_) [Yarn](https://yarnpkg.com/) installed on your machine
```

> - (Optional) [Docker Compose](https://docs.docker.com/compose/) installed on your machine

Reviewer 2: ```suggestion
- (_optional_) [Docker Compose](https://docs.docker.com/compose/) installed on your machine
```

:::

## Development and/or Staging environments

For working with Strapi locally on your host machine you can use the [Dockerfile](https://docs.docker.com/engine/reference/builder/), and if needed the [docker-compose.yml](https://docs.docker.com/compose/compose-file/) can also be used to start up a database container.

> Both methods require an existing Strapi project or a new one created with [Strapi Quick Start](/developer-docs/latest/getting-started/quick-start.md).

Reviewer 2: ```suggestion
Both methods require an existing Strapi project or a new one created (see [Quick Start](/developer-docs/latest/getting-started/quick-start.md) guide).
```

      DATABASE_PASSWORD: ${DATABASE_PASSWORD}
      JWT_SECRET: ${JWT_SECRET}
      ADMIN_JWT_SECRET: ${ADMIN_JWT_SECRET}
      APP_KEYS: ${APP_KEYS}
      NODE_ENV: ${NODE_ENV}
    volumes:
      - ./config:/opt/app/config
      - ./src:/opt/app/src
      - ./package.json:/opt/package.json
      - ./yarn.lock:/opt/yarn.lock
      - ./.env:/opt/app/.env
      - ./public/uploads:/opt/app/public/uploads
    ports:
      - '1337:1337'
    networks:
      - strapi
    depends_on:
      - strapiDB
      
  strapiDB:
    container_name: strapiDB
    platform: linux/amd64 #for platform error on Apple M1 chips
    restart: unless-stopped
    env_file: .env
    image: postgres:12.0-alpine
    environment:
      POSTGRES_USER: ${DATABASE_USERNAME}
      POSTGRES_PASSWORD: ${DATABASE_PASSWORD}
      POSTGRES_DB: ${DATABASE_NAME}
    volumes:
      - strapi-data:/var/lib/postgresql/data/ #using a volume
      #- ./data:/var/lib/postgresql/data/ # if you want to use a bind folder

    ports:
      - '5432:5432'
    networks:
      - strapi

volumes:
  strapi-data:

networks:
  strapi:
    name: Strapi
    driver: bridge
```

</code-block>

</code-group>

> ## Production Environments

Reviewer 2: It would be useful to have 1 or 2 introductory sentences here, quickly explaining/summing up what are the differences between development/staging and production environments and why some steps are optional.

Also, since we don't have a secondary TOC on VuePress and h3s are not currently displayed in the TOC, it would be nice to give quick links to the sub-sections.

For instance, something like:
"Using is a Docker image in production is different because [blah blah] and [blah blah]. 

Once the [Dockerfile](#dockerfile) is created, the [production container](#link-to-relevant-section) can be built. Optionally, the container can be published to a [registry](#link-to-subsection). [Community tools](#) can help you in the process."

Depending on your hosting provider you may need to use a different command to publish your image. It is recommended you review the [Docker documentation](https://docs.docker.com/engine/reference/commandline/push/) for more information on publishing Docker images with more advanced options.

Some popular hosting providers are:

- [AWS ECR](https://aws.amazon.com/ecr/)
- [Azure Container Registry](https://azure.microsoft.com/en-us/services/container-registry/)
- [GCP Container Registry](https://cloud.google.com/container-registry)
- [Digital Ocean Container Registry](https://www.digitalocean.com/products/container-registry/)
- [IBM Cloud Container Registry](https://www.ibm.com/cloud/container-registry)
- [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)
- [Gitlab Container Registry](https://docs.gitlab.com/ee/user/packages/container_registry/)

## Community Tools

There are several community tools that can be used to deploy Strapi to various cloud providers and can assist you in setting up Docker both in a development and production environment.

> We strongly support our community efforts and encourage you to check out the following tools, please do help support them by contributing to their development.

Reviewer: ```suggestion
We strongly support our community efforts and encourage you to check out the following tools, please help support them by contributing to their development.
```

If you would like to add your tool to this list, please open a pull request on the [Strapi documentation repository](https://github.com/strapi/documentation).

### @strapi-community/dockerize

The `@strapi-community/dockerize` package is a CLI tool that can be used to generate a Dockerfile and docker-compose.yml file for a Strapi project.

> To get started you simply need to run `npx @strapi-community/dockerize@latest` within an existing Strapi project folder and follow the CLI prompts.

Reviewer: ```suggestion
To get started run `npx @strapi-community/dockerize@latest` within an existing Strapi project folder and follow the CLI prompts.
```

For more information please see the official [GitHub repository](https://github.com/strapi-community/strapi-tool-dockerize) or the [npm package](https://www.npmjs.com/package/@strapi-community/dockerize).

<!-- TODO: Wait for community to finish up on some things before adding the below section. -->
<!-- ### @strapi-community/deployify -->

> ## Docker FAQ

Reviewer 2: I know in the past we've had many sections having their own FAQ but I'm not a big supporter of it, I'd rather have these moved to the general FAQ (https://docs.strapi.io/developer-docs/latest/getting-started/troubleshooting.html). What do you think?

Author: Not in this case Piwi please, these are very specific to Docker and they are questions we absolutely will be getting very often.

I want them as easy as possible to find so I don't have to copy/paste more than one link to a user when they mention docker.

Reviewer 2: Do you plan to add more entries to the Docker FAQ?
If not, I had a second look at the content and think that they can be moved to the introduction. 

The first point is basically repeating what we mention at the top of the document, and the second one is actually the answer to a question I asked in a comment in this PR (production vs. staging images) 😄 . Could probably be explained in the introduction, with less words.
Just a suggestion but that would simplify and lighten this document a bit.
Then you could use a single link, the page's link, for any question. What do you think?
